docs/2022-07-18-JSON-RPC-Server-API.md

```diff
@@ -388,6 +388,127 @@ Same as for execute
+
+#### Implication is **valid**
+
+A constrained substitution as the result, and this is only returned if the implication is valid.
+
+```
+{
+  "jsonrpc": "2.0",
+  "id": 1,
+  "result": {
+    "validity": {"tag": "implication-valid"},
+    "substitution": {"format": "KORE", "version": 1, "term": {}},
+  }
+}
+```
+
+#### Implication **invalid**: terms do not match
+
+Matching between antecedent and consequent configurations failed (different constructors, shared variables, sort error, etc.), constraints has not been subsumption been attempted. No matching substitution is returned.
+
+```
+{
+  "jsonrpc": "2.0",
+  "id": 1,
+  "result": {
+    "validity": {"tag": "implication-valid",
+                 "contents": {"tag": "matching-failed",
+                              "contents": "Shared variables: X:SortWordStack{}"
+                 }},
+  }
+}
+```
+
+#### Implication **invalid**: terms match, but constraints subsumption failed
+
+Matching between antecedent and consequent configurations is successful, but constraints do not agree. Response contains the matching substitution and the unsatisfiable core of constraints.
+
+```
+{
+  "jsonrpc": "2.0",
+  "id": 1,
+  "result": {
+    "validity": {"tag": "implication-valid",
+                 "contents": {"tag": "constraint-subsumption-failed",
+                              "contents": {"format": "KORE", "version": 1, "term": {}}
+                 }},
+    "substitution": {"format": "KORE", "version": 1, "term": {}},
+  }
+}
+```
+
+#### Implication **unknown**: matching indeterminate
+
+The matching algorithm is incomplete and may return an indeterminate result. The response will contains the subterms that the algorithm could not know how to match. No substitution is returned.
+
+```
+{
+  "jsonrpc": "2.0",
+  "id": 1,
+  "result": {
+    "validity": {"tag": "implication-unknown",
+                 "contents": {"tag": "matching-unknown",
+                              "contents": {"first" : {"format": "KORE", "version": 1, "term": {}}
+                                          ,"second" : {"format": "KORE", "version": 1, "term": {}}
+                                          }
+                 }},
+  }
+}
+```
+
+#### Implication **unknown**: constraint subsumption indeterminate
+
+If matching is successful, but the constraint solver procedure (internal, or the SMT solver if enabled) returned "unknown". Response contains the matching substitution and the unknown core of constraints.
+
+```
+{
+  "jsonrpc": "2.0",
+  "id": 1,
+  "result": {
+    "validity": {"tag": "implication-unknown",
+                 "contents": {"tag": "constraint-subsumption-unknown",
+                              "contents": {"tag": "constraint-subsumption-failed",
+                                           "contents": {"format": "KORE", "version": 1, "term": {}}
+                 }},
+    "substitution": {"format": "KORE", "version": 1, "term": {}},
+  }
+}
+```
```

Comment:
How is this different than "Implication **invalid**: terms match, but constraints subsumption failed"

Reply:
I thought, we'd want to have a tri-state constraint subsumption check, just like we have a tri-stat matching routine. But perhaps we can indeed expose the "failed" and "unknown" constraint subsumption as just "failed" and have less user-facing cases.

Reply:
Part of the confusion for me is that for the failure cases, you're using "terms don't `match`" as the criteria, but that's not how the tri-state matching algorithm works. It either reports back:

- Terms _match_ with substitution _alpha_.
- Terms _definitely don't unify_.
- Unknown.

The key here is that unification and matching are different algorithms, different complexities, and different properties. Saying a term matches another means it's _completely subsumed_, and saying that they _don't unify_ means they have _no overlap_.

So I guess the language here needs to be cleared up, because here for example 
![image](https://github.com/runtimeverification/haskell-backend/assets/12721972/bc953908-42a5-46b0-bba1-19b96039d2d5)
 we're saying "terms don't match" because of differing constructor, where I think it should be saying "terms don't unify".

Unification and matching are not interchangeable. See https://github.com/runtimeverification/hs-backend-booster/issues/193 again for the desired interface.